### ISSUE TITLE
After upgrading to 8.1 and 8.1.1 and 8.1.3 a directory is shown as file in web view and no longer synced by clients

### ISSUE THREAD

**Issue by norgoe** [owncloud/core#17881]
Hi,

I have a directory t530.conf storing about 1,1GB of files which was shown correctly in the files pane in owncloud 8.0.
In 8.1 it is shown as a text file with a size of 1.1GB. If I click the editor for textfiles is opened with an empty file. If I initiate a download via the file view I get a large zip file containing 1.1GB of files.

**Comment by karlitschek** [comment]
@icewind1991 Any idea?

**Comment by norgoe** [comment]
Update:
The solution was to rescan all files of the user on the server. This proved to be harder as it sounds as meanwhile the desktop sync client deleted almost all (but not all!) content of the t530.conf folder on my client. As this folder holds config files for my windows manager, too, I had a really weird reacting i3 window manager.
After rescanning the user files the directory is browsable via web gui again and the owncloud client (after restarting it) synced the content to my hdd again.

Maybe all this was induced by other problems I had with the update.
First no files where shown (disabling the picture plugin solved this).
Than I reactivated calendar plugin which did not update and effectively killed occ. After resolving this issue I found the folder problems.

**Comment by PVince81** [comment]
@norgoe do you have any traces of the problem ? If you managed to fix them then it will be close to impossible to analyze/fix it.

One theory is that maybe the mime type entry in oc_filecache for that folder got corrupt for some mysterious reasons. Running a rescan fixed it.

**Comment by DeepDiver1975** [comment]
I guess we should move the instance into maintenance mode while running the rescan to prevent any client to cause issues on the file storage?

**Comment by norgoe** [comment]
@PVince81 
at least I did not initate any tracing so if it is not logged automatically there will be no trace.
My main goal was to get my data back which is much harder if your desktop environment does not react like expected cause config files got deleted ;)

Maybe it is related to messages like the following which are still spamming my owncloud.log since I updated to 8.1:

{"reqId":"SOMEOBSCUREID","remoteAddr":"xxx.xxx.xxx.xxx","app":"PHP","messag
e":"Declaration of OCA\\Files_Sharing\\External\\Scanner::scanFile() should be c
ompatible with OC\\Files\\Cache\\Scanner::scanFile($file, $reuseExisting = 0, $p
arentId = -1, $cacheData = NULL, $lock = true) at \/var\/www\/owncloud\/apps\/fi
les_sharing\/lib\/external\/scanner.php#121","level":3,"time":"2015-07-28T10:34:
48+00:00"}

@DeepDiver1975 
I did this before issuing the rescan by hand.

**Comment by PVince81** [comment]
@DeepDiver1975 could be an idea, but I'm not sure. @icewind1991 what do you think ?
Note that there is also a cron job for background scan that scans in a more slowly manner, but could have the same side effects. I'm not sure if maintenance mode is the best solution here.

It should work better with transactional file locking, which also has scanner locking.

**Comment by norgoe** [comment]
OK,
the same shit just happened again after upgrading from 8.1 to 8.1.1.
The same folder with all it's content is shown as an empty text file but with 1,1gb size.
The owncloud client does not see the folder at all and deletes it locally.

I am snapshotting the data and the owncloud partition now.
Afterwards I will rescan all files again like last time to get a wowkring installation back.

How can I make the forementioned traces to help debugging the problem?

**Comment by norgoe** [comment]
Ok folks, it is getting really annoying.
After noone was interested in the problem I chose to start from scratch.
I installed a new xen vm (debian jessie) and installed owncloud 8.1.1 from deb http://download.opensuse.org/repositories/isv:/ownCloud:/community/Debian_8.0/ /
I setup a new owncloud 8.1.1 instance on mysql without using any backups (except my calendar and contacts as ics/vcard).
I than synced (with a completely wiped owncloud config) my ubuntu laptop's data again with the "t530.conf".
Everythin worked well until today when I updated owncloud from 8.1.1 to 8.1.3 using
apt-get update
apt-get dist-upgrade

Now again the folder is shown as a file (this time 479MB as I deleted some stuff in the directory) in the webgui while the "file" attributes on the server clearly state that it is a directory:
drwxr-xr-x   11 www-data www-data      4096 Sep  3 00:49 t530.conf


As this happens on a newly setup server there has to be a nasty bug in the server update process.
![folder shown as file](https://cloud.githubusercontent.com/assets/6221132/9903755/30a3763a-5c79-11e5-9018-832882f6c344.png)

**Comment by icewind1991** [comment]
Can't reproduce.

It seems like something is messing with `is_dir` on the server, what os, storage backend is used?

**Comment by norgoe** [comment]
It's a xen vm hosting debian jessie (formerly debian wheezy) and using a separate ext4 partition for the user-data handled by the vm itself.

**Comment by PVince81** [comment]
Cannot reproduce on 8.1.3.

Did you disable all third party apps ?
Did you manually create symbolic links or hard links inside the data folder ? (just in case)
Do you have any files in that folder ? Does it happen also if the folder is empty ?

To check if PHP is behaving strangely, you could try writing a quick php script:
```php
   <?php
   print(is_dir('data/user1/files/t530.conf'));
```

Put it in your "owncloud" folder then open it from the web browser.
It should return "1".

This is assuming that you are not using the ownCloud app "External Storage". (this is what @icewind1991 meant with storage backend)

**Comment by norgoe** [comment]
Did you disable all third party apps ?
- During or before the update? No
- The update process itself disabled them all and I hade to manually reanable all of them.

Did you manually create symbolic links or hard links inside the data folder ? (just in case)
- No, I try to mess around as little as possible on the system. The only thing I did was rescanning to get my laptops config files to sync again so I do have a working x session after the next reboot.

Do you have any files in that folder ? Does it happen also if the folder is empty ?
- One file and a bunch of directories which contain files. Summing up at roughly 480MB.
- At least I have not recognized it on any other folder (empty or not).

The result of the php check is 1 after changing the path corresponding to the owncloud config:
  'datadirectory' => '/mnt/owncloud-data/user-data',

And no, I am not using "External Storage". I am using a LVM Partition handed from the Dom0 to the DomU.

**Comment by PVince81** [comment]
Hmm this doesn't make any sense.

I wonder if any part of the code is trying to re-detect the mime type and is using the ".conf" believing it's a text file instead of a directory... But looking at the code it seems to be correct: https://github.com/owncloud/core/blob/v8.1.1/lib/private/files/storage/common.php#L224
It first tries to detect if it's a dir, and only if it isn't it will resort to mime type detection based on the extension.

@norgoe if possible, you could check the "oc_filecache" entry before and after the update:
`select f.*,m.* from oc_filecache f, oc_mimetypes m where path like '%/t530.conf' and f.mimetype=m.id;`.

If the mimetype changed after the update, it means that there's something in the update code that messes it up.

**Comment by PVince81** [comment]
Oohhh, this reminds me that there's a repair step that does stuff with mime types, and look what I found: https://github.com/owncloud/core/blob/v8.1.1/lib/repair/repairmimetypes.php#L226

Now let's see if this can be reproduced when the conditions triggering that repair step are met...

**Comment by PVince81** [comment]
Steps:

1. Setup 8.0.8
1. Create a folder "test.conf" in the web UI
1. Upload some files into the folder "test.conf"
1. Upgrade to 8.1.1

And here I see the mimetype has changed...

The repair step that fixes mimetypes is probably not checking if it's a folder...

**Comment by norgoe** [comment]
Thanks for finding ther error. At least I did not repeat a terrible mistake three times in a row.

But if my mind is not playing games on me this just happened to the folder name t530.conf but not to another folder named x200t.conf hosted on the same owncloud account.
But it is also possible that I rescanned the files before using the seldom used thinkpad x200t again and thus did not recognize that the x200t.conf folder was "gone", too.
